Sources/TuistLoader/Models+ManifestMappers/Headers+ManifestMapper.swift

```diff
@@ -11,59 +11,65 @@ extension TuistGraph.Headers {
-                let resolved = try generatorPaths.resolve(path: path).pathString
-                let absolute = AbsolutePath(resolved)
-                let globs = AbsolutePath(absolute.dirname).glob(absolute.basename)
-                result.formUnion(globs)
-            }
-            return result
-        }
-
-        func unfoldGlob(_ path: AbsolutePath, excluding: Set<AbsolutePath>) -> [AbsolutePath] {
-            FileHandler.shared.glob(AbsolutePath.root, glob: String(path.pathString.dropFirst())).filter {
-                guard let fileExtension = $0.extension else {
-                    return false
-                }
-                return TuistGraph.Headers.extensions.contains(".\(fileExtension)") && !excluding.contains($0)
-            }
+    static func from(
+        manifest: ProjectDescription.Headers,
+        generatorPaths: GeneratorPaths,
+        productName: String?
+    ) throws -> TuistGraph.Headers {
+        let resolvedUmbrellaPath = try manifest.umbrellaHeader.map { try generatorPaths.resolve(path: $0) }
+        let headersFromUmbrella = try resolvedUmbrellaPath.map {
+            Set(try UmbrellaHeaderHeadersExtractor.headers(from: $0, for: productName))
         }
 
         var autoExlcudedPaths = Set<AbsolutePath>()
         let publicHeaders: [AbsolutePath]
         let privateHeaders: [AbsolutePath]
         let projectHeaders: [AbsolutePath]
 
-        func resolveHeaders(_ list: FileList?) throws -> [AbsolutePath] {
+        let allowedExtensions = TuistGraph.Headers.extensions
+        func unfold(_ list: FileList?,
+                    isPublic: Bool = false) throws -> [AbsolutePath]
+        {
             guard let list = list else { return [] }
-            return try list.globs.flatMap {
-                unfoldGlob(
-                    try generatorPaths.resolve(path: $0.glob),
-                    excluding: (try resolveExcluding($0.excluding)).union(autoExlcudedPaths)
-                )
+            var result = try list.globs.flatMap {
+                try $0.unfold(generatorPaths: generatorPaths) { path in
+                    guard let fileExtension = path.extension,
+                          allowedExtensions.contains(".\(fileExtension)"),
+                          !autoExlcudedPaths.contains(path)
+                    else {
+                        return false
+                    }
+                    if isPublic, let headersFromUmbrella = headersFromUmbrella {
+                        return headersFromUmbrella.contains(path.basename)
+                    }
+                    return true
+                }
+            }
+            // be sure, that umbrella was not added before
+            if let resolvedUmbrellaPath = resolvedUmbrellaPath,
+               isPublic, // to use this logic only when public is used
+               !result.contains(resolvedUmbrellaPath)
+            {
+                result.append(resolvedUmbrellaPath)
             }
+            return result
```

Comment:
minor: This can be done outside the `unfold` method to simplify it _(right now quite a lot is happening here)_